(Yet Another) Memory Blocks Allocator

dcpleung opened the pull request:
This introduces (yet another) memory blocks allocator. This allows memory blocks to be dynamically allocated from a designated memory region, where:

* All memory blocks have a single fixed size.
* Multiple blocks can be allocated or freed at the same time.
* A group of blocks allocated together may not be contiguous. This is useful for operations such as scatter-gather DMA transfers.
* Bookkeeping of allocated blocks is done outside of the associated buffer (unlike memory slab). This allows the buffer to reside in memory regions where these can be powered down to conserve energy.

This also adds wrapper interface similar to `sys_multi_heap` so multiple allocators can be handled as a group.

dcpleung committed:
lib: os: introduce memory blocks allocator

This introduces yet another memory blocks allocator where:

() All memory blocks have a single fixed size.

() Multiple blocks can be allocated or freed at the same time.

() A group of blocks allocated together may not be contiguous.
   This is useful for operations such as scatter-gather DMA
   transfers.

() Bookkeeping of allocated blocks is done outside of
   the associated buffer (unlike memory slab). This allows
   the buffer to reside in memory regions where these can be
   powered down to conserve energy.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
doc: add a page for the new memory blocks allocator

This adds a page for the new memory blocks allocator,
describing what it is and how to use it.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
tests: add new tests for the new memory blocks allocator

This adds some new tests to make sure the new memory blocks
allocator work correctly.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
lib: os: mem_blocks: add memory blocks allocator groups

This adds similar ability of sys_multi_heap to the memory blocks
allocator, where a choice function can be used to select
which allocator (of a group) is used for memory block allocation.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
tests: lib/mem_blocks: add tests for multi mem_blocks allocator

This adds a few tests to make sure the multi memory blocks
allocator functions work correctly.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
Compliance check failure is due to new typedefs.

teburd commented on lib/os/mem_blocks.c:
```diff
@@ -0,0 +1,246 @@
+		goto out;
+	}
+
+	for (i = 0; i < count; i++) {
```
This loop with the following free could I suppose potentially lead to some allocation races where two cores contend on allocating some number of blocks, and neither succeeds to allocate enough blocks despite enough being there for either of the calls... Nothing necessarily bad comes from it, but maybe not the most friendly outcome if that's a common use case

teburd reviewed (approved):
It's a nice way of handling some potentially gnarly slab like allocation scenarios that just can't be covered by the current slab API. For that it does its job well I think! One comment is really a potential question about how this is possibly used.

SebastianBoe commented on lib/os/CMakeLists.txt:
```diff
@@ -24,6 +24,7 @@ zephyr_sources(
   heap-validate.c
   bitarray.c
   multi_heap.c
+  mem_blocks.c
```
If this PR is accepted I would like to see compilation of mem_blocks.c be opt-in and dependent on Kconfig to save build-time for all the builds that don't use it.

I know adding it unconditionally follows prior-art, but we can't fix prior-art without making breaking changes. But for new features we can do this the right way.

dcpleung replied:
Added kconfig for the API.

dcpleung committed:
lib: os: group heap related kconfigs under a menu

This groups the heap related kconfigs under a menu
so they are under a group (!).

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

SebastianBoe reviewed (approved):
CMake and KConfig is OK

nashif commented:
@lgirdwood  can you please take a look.

npitre commented:
A few comments:

- Communicating the list of blocks in an uintptr_t array is confusing.
  One could think they are offsets into the storage area (I had to look at the
  code to be sure). You should use an array of void pointers instead to make it
  obvious.

- You should avoid `POINTER_TO_UINT()` and its reverse twin. An int typically
  can't hold a pointer on 64-byts systems. I know the implementation casts to
  uintptr_t and it is OK here. But it makes the code clearer (and shorter)
  to perform the cast directly.

- You get uintptr_t in `sys_mem_blocks_free()`, convert them into pointers for
  `free_one()`, and then it is reconverted to uintptr_t. Addressing both points
  above would remove much of that back and forth. It is also fine to do
  arithmetic on an uint8_t pointer and assign it to a void* variable without
  any casts.

- You should consider storing the block size as a shift value. That would
  avoid relying on a divide operation which is quite slow on some MCUs.
  Linux has an `ilog2()` that resolves to a constant at compile time when the
  argument is constant. Having the same for Zephyr would be quite useful here
  and elsewhere too.

dcpleung committed:
math: introduce macro for integer log2

This introduces macro for calculating integer log2.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
> A few comments: [...]

Updated to address these comments.

nashif commented:
please rebase

dcpleung committed:
lib: os: mem_blocks: add alloc/free event notifications

This adds event notification for mem_blocks using the new
heap listener interface.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
tests: lib/mem_blocks: add test for heap listeners

This adds simple tests to make sure the heap listener interface
works with memory blocks allocator.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
Added 2 commits to support alloc/free event notifications via heap listener interface.
